### PR TITLE
feat: bump node version to 18.17.1

### PR DESCRIPTION
**Pull request by pac-guerreiro**
### Details

Onyx currently enforces developers to use NodeJS 16.15.1. [As of September, NodeJS 16.x will reach its End-of-life (EOL)](https://github.com/nodejs/release#release-schedule), meaning there will be no more updates, patches, or security fixes. Additionally, https://github.com/reactwg/react-native-releases/discussions/64.

Using an outdated version of NodeJS not only exposes the application to potential security vulnerabilities, compatibility issues, and prevents developers from leveraging the latest features, optimizations, and improvements available in newer versions of NodeJS.

### Related Issues

- https://github.com/Expensify/App/issues/25824

### Automated Tests

Not necessary since this PR only changes project configuration files.

### Linked PRs

None.

### CODE REVIEW

**Review by bondydaa — DISMISSED** [review]
👍 seems right to me but going to assign some others who are more familiar as well.

File: package.json
Comment (fabioh8010) [inline package.json:102]:
```suggestion
    "node": ">=16.15.1 <=18.17.1",
    "npm": ">=8.11.0 <=9.6.7"
```

@pac-guerreiro I think that, for now, we have to specify a range of supported NodeJS versions. I'm thinking about the following situation:

1. This PR is merged and available as a new version to use on E/App
2. Developer runs `npm install` to update dependencies
3. They got an error because `react-native-onyx` is expecting `18.17.1` but E/App currently forces use to use `16.15.1` until we change there as well

After both repos are updated then we can lock this on `18.17.1` only. WDYT?

**Comment by roryabraham** [comment]
I think this should be subject to the [merge freeze](https://expensify.slack.com/archives/C01GTK53T8Q/p1693347701566439), putting it on HOLD. @pac-guerreiro please take this off HOLD when the merge freeze is lifted

**Review by bondydaa — DISMISSED** [review]
👍 again, i think feature freeze is over so taking off hold but @roryabraham you want to double check this and merge if you feel 👍 about it?

**Review by roryabraham — APPROVED** [review]
LGTM as long as we follow-up to remove the range and pin to 18

**Comment by roryabraham** [comment]
However, I can't merge this PR due to it containing unsigned commits. @pac-guerreiro, please set up GPG signing for your commits and register your GPG key with GitHub, then rebase this PR, reword each commit, and force-push.

image

**Review by roryabraham — DISMISSED** [review]

**Comment by roryabraham** [comment]
Tests are failing

**Comment by pac-guerreiro** [comment]
@roryabraham 

This test is failing but it's unrelated to my changes:

Screenshot 2023-09-21 at 14 23 24

**Comment by roryabraham** [comment]
@pac-guerreiro tests appear to be passing on main, please try merging `main` into this branch?

**Comment by pac-guerreiro** [comment]
@roryabraham 

I just rebased the branch with main but the issue persists, although locally the tests are passing. Maybe this is a cache issue? Could you re-run the workflow and maybe delete the cache?

**Comment by roryabraham** [comment]
Deleted caches and re-running

**Comment by pac-guerreiro** [comment]
@roryabraham 

I think I might have found why tests are failing:

Screenshot 2023-10-02 at 15 53 00
Screenshot 2023-10-02 at 15 53 38

It seems that we're using an old version of `actions/setup-node@v1` and I'm trying to use a feature to read the node version from a config file which is only available in `actions/setup-node@v3`

**Comment by AndrewGable** [comment]
Can you update it in this PR and see if the tests pass?

**Comment by roryabraham** [comment]
For setup-node feel free to pin it directly to the @v3 tag instead of a commit hash

**Comment by pac-guerreiro** [comment]
@roryabraham I'm having issues with GPG. Somehow I'm unable to commit anything, even after generating a new key 😢

**Comment by pac-guerreiro** [comment]
@AndrewGable @roryabraham All checks have passed! 😄

**Review by AndrewGable — APPROVED** [review]

**Review by bondydaa — APPROVED** [review]

**Comment by bondydaa** [comment]
going to merge this so it doesn't get tripped up by some other conflict, looks like rory had previously approved it but we didn't merge b/c of failing tests that are now fixed.

**Pull request merged**
